avaframe/com1DFA/DFAfunctionsCython.pyx

```diff
@@ -741,32 +770,55 @@ def updatePositionC(cfg, particles, dem, force, DT):
   log.debug('Entrained DFA mass: %s kg', np.asarray(massEntrained))
   particles['kineticEne'] = TotkinEneNew
   particles['potentialEne'] = TotpotEneNew
-  if peakKinEne < TotkinEneNew:
-    particles['peakKinEne'] = TotkinEneNew
-  if particles['peakMassFlowing'] < massFlowing:
-    particles['peakMassFlowing'] = massFlowing
-
-  if cfg['stopCritType'] == 'massFlowing':
-    value = massFlowing
-    peakValue = particles['peakMassFlowing']
-    stop = (value <= stopCrit*peakValue) and peakValue > 0
-  elif cfg['stopCritType'] == 'kinEnergy':
-    value = TotkinEneNew
-    peakValue = particles['peakKinEne']
-    stop = value <= stopCrit*peakValue
+
+  if typeStop == 1:
+    # typeStop = 1 for initialisation step where particles are redistributed to reduce SPH force
+    # here stop criterion based on SPHForce
+    value = totForceSPHNew
+    peakValue = particles['peakForceSPH']
+    oldValue = particles['forceSPHIni']
+    particles['forceSPHIni'] = totForceSPHNew
+    if oldValue == 0.0:
+      stop = False
+    elif value < 1.:
+      stop = oldValue/value < 1.001
+    else:
+      stop = value <= stopCritIni*peakValue
+      log.info('SPHFORCE value %f and stop value %f' % (totForceSPHNew, stopCritIni*peakValue))
+    if peakForceSPH < totForceSPHNew:
+      particles['peakForceSPH'] = totForceSPHNew
   else:
-    message = 'stopCritType %s is not defined' % cfg['stopCritType']
-    log.error(message)
-    raise AssertionError(message)
+    # avalanche computation stop criterion based on kinetic energy or massFlowing
+    if peakKinEne < TotkinEneNew:
+      particles['peakKinEne'] = TotkinEneNew
+    if particles['peakMassFlowing'] < massFlowing:
+      particles['peakMassFlowing'] = massFlowing
+
+    if cfg['stopCritType'] == 'massFlowing':
+      value = massFlowing
+      peakValue = particles['peakMassFlowing']
+      stop = (value <= stopCrit*peakValue) and peakValue > 0
```

Comment:
This part is not covered, but is crucial. please add to a pytest

Reply:
reference in a new issue